annotatedcsv/decode.go

```diff
@@ -0,0 +1,463 @@
+//
+// When it's a pointer to a slice, the slice is changed
+// to have one element for each column (reusing
+// space in the slice if possible), and each element is
+// set to the value in the corresponding column.
+//
+// When decoding into an empty interface value, the resulting
+// type depends on the column type:
+//
+// - string, tag or unrecognized: string
+// - double: float64
+// - unsignedLong: uint64
+// - long: int64
+// - boolean: bool
+// - duration: time.Duration
+// - dateTime: time.Time
+//
+// Any value can be decoded into a string without
+// error - the result is the value in the CSV, so
+//
+//     var row []string
+//     r.Decode(&row)
+//
+// will always succeed and provide all the values in the column as strings.
+//
+// Similarly, any row can be decoded into []interface{}, which will convert
+// all values into elements of the slice according to the rules for empty interfaces above.
+//
+// Slice capacity will be reused when possible.
+func (r *Reader) Decode(x interface{}) error {
+	t := reflect.TypeOf(x)
+	if err := r.initColumns(t, r.cols); err != nil {
+		return err
+	}
+	return r.decodeRow(reflect.ValueOf(x), r.row)
+}
+
+// initColumns initializes r.colSetters for the given columns.
+// It's called once per section.
+func (r *Reader) initColumns(t reflect.Type, columns []Column) error {
+	if t == r.decodeType {
+		return nil
+	}
+	if t.Kind() != reflect.Ptr {
+		return fmt.Errorf("cannot decode into non-pointer type %v", t)
+	}
+	et := t.Elem()
+	if et.Kind() != reflect.Struct && et.Kind() != reflect.Slice {
+		return fmt.Errorf("cannot decode into %v", t)
+	}
+	setters := make([]fieldSetter, len(columns))
+	switch et.Kind() {
+	case reflect.Struct:
+		fieldsMap := make(map[string]reflect.StructField)
+		fields := ireflect.VisibleFields(et)
+		for _, f := range fields {
+			name := f.Name
+			if tag, ok := f.Tag.Lookup("csv"); ok {
+				switch {
+				case tag == "-":
+					continue
+				case tag == "-,":
+					tag = "-"
+				case strings.Contains(tag, ","):
+					i := strings.Index(tag, ",")
+					if i == len(tag)-1 {
+						tag = tag[:len(tag)-1]
+					} else {
+						return fmt.Errorf("tag attributes are not supported")
+					}
+				}
+				name = tag
+			}
+			fieldsMap[name] = f
+		}
+
+		for i, col := range columns {
+			f, ok := fieldsMap[col.Name]
+			if !ok {
+				continue
+			}
+			ftype := f.Type
+			colType, ok := columnTypes[col.Type]
+			if !ok {
+				// ignore invalid type and use string
+				colType = stringCol
+			}
+			fkind := fieldKindOf(ftype)
+			convert, ok := conversions[conv{colType, fkind}]
+			if !ok {
+				return fmt.Errorf("cannot convert from column type %s to %v", col.Type, ftype)
+			}
+			setters[i] = func(v reflect.Value, colIndex int) error {
+				if err := r.convertColumnValue(v.FieldByIndex(f.Index), colIndex, convert); err != nil {
+					return err
+				}
+				return nil
+			}
+		}
+		r.decodeRow = func(v reflect.Value, row []string) error {
+			v = v.Elem()
```

Comment:
```suggestion
			if v.IsNil() {
			    return fmt.Errorf("decode into nil %s", v.Type())
			}
			v = v.Elem()
```

or maybe a panic. Anyway, perhaps nicer to have an explicit message rather than a panic inside reflect.